Mathlib/Algebra/Group/Hom/Instances.lean

```diff
@@ -19,56 +22,74 @@ operations.
 @[to_additive "`(M →+ N)` is an `AddCommMonoid` if `N` is commutative."]
-instance MonoidHom.instCommMonoid [MulOneClass M] [CommMonoid N] :
-    CommMonoid (M →* N) where
-  mul := (· * ·)
-  mul_assoc := by intros; ext; apply mul_assoc
-  one := 1
-  one_mul := by intros; ext; apply one_mul
-  mul_one := by intros; ext; apply mul_one
-  mul_comm := by intros; ext; apply mul_comm
-  npow n f :=
-    { toFun := fun x => f x ^ n, map_one' := by simp, map_mul' := fun x y => by simp [mul_pow] }
-  npow_zero f := by
-    ext x
-    simp
-  npow_succ n f := by
-    ext x
-    simp [pow_succ]
+instance MonoidHom.instCommMonoid [MulOneClass M] [CommMonoid N] : CommMonoid (M →* N) :=
+  fast_instance%
+    DFunLike.coe_injective.commMonoid DFunLike.coe rfl (fun _ _ => rfl) (fun _ _ => rfl)
+
+instance AddMonoidHom.instIntSMul [AddZeroClass M] [AddCommGroup N] : SMul ℤ (M →+ N) where
+  smul a f :=
+    { toFun := a • f
+      map_zero' := by simp [zsmul_zero]
+      map_add' x y := by simp [zsmul_add] }
+
+@[to_additive existing AddMonoidHom.instIntSMul]
+instance MonoidHom.instIntPow [MulOneClass M] [CommGroup N] : Pow (M →* N) ℤ where
+  pow f n :=
+    { toFun := f ^ n
+      map_one' := by simp
+      map_mul' x y := by simp [mul_zpow] }
+
+@[to_additive (attr := simp)]
+lemma MonoidHom.zpow_apply [MulOneClass M] [CommGroup N] (f : M →* N) (z : ℤ) (x : M) :
+    (f ^ z) x = (f x) ^ z :=
+  rfl
 
 /-- If `G` is a commutative group, then `M →* G` is a commutative group too. -/
 @[to_additive "If `G` is an additive commutative group, then `M →+ G` is an additive commutative
       group too."]
-instance MonoidHom.instCommGroup {M G} [MulOneClass M] [CommGroup G] : CommGroup (M →* G) :=
-  { MonoidHom.instCommMonoid with
-    inv := Inv.inv,
-    div := Div.div,
-    div_eq_mul_inv := by
-      intros
-      ext
-      apply div_eq_mul_inv,
-    inv_mul_cancel := by intros; ext; apply inv_mul_cancel,
-    zpow := fun n f =>
-      { toFun := fun x => f x ^ n,
-        map_one' := by simp,
-        map_mul' := fun x y => by simp [mul_zpow] },
-    zpow_zero' := fun f => by
-      ext x
-      simp,
-    zpow_succ' := fun n f => by
-      ext x
-      simp [zpow_add_one],
-    zpow_neg' := fun n f => by
-      ext x
-      simp [zpow_natCast, -Int.natCast_add] }
+instance MonoidHom.instCommGroup [MulOneClass M] [CommGroup N] : CommGroup (M →* N) :=
+  fast_instance%
+    DFunLike.coe_injective.commGroup DFunLike.coe
+      rfl (fun _ _ => rfl) (fun _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl)
+
+@[to_additive]
+instance [MulOneClass M] [CommMonoid N] [IsLeftCancelMul N] : IsLeftCancelMul (M →* N) where
+  mul_left_cancel _ _ _ h := DFunLike.coe_injective <| mul_left_cancel (congrArg DFunLike.coe h)
+
+@[to_additive]
+instance [MulOneClass M] [CommMonoid N] [IsRightCancelMul N] : IsRightCancelMul (M →* N) where
+  mul_right_cancel _ _ _ h := DFunLike.coe_injective <| mul_right_cancel (congrArg DFunLike.coe h)
```

Comment:
We already have `Function.Injective.isLeftCancelMulZero`. Should we acquire `Function.Injective.isLeftCancelMul`?